idaes/models_extra/power_generation/unit_models/soc_submodels.py

```diff
@@ -0,0 +1,3997 @@
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+
+import copy
+import enum
+
+from pyomo.common.config import ConfigBlock, ConfigValue, In
+from pyomo.dae import DerivativeVar
+import pyomo.environ as pyo
+from pyomo.network import Port
+
+
+from idaes.core import declare_process_block_class, UnitModelBlockData, useDefault
+from idaes.models.unit_models.heat_exchanger import HeatExchangerFlowPattern
+import idaes.core.util.scaling as iscale
+from idaes.core.util.exceptions import ConfigurationError, InitializationError
+from idaes.core.util.math import safe_log
+from idaes.core.util import get_solver
+import idaes.core.util.model_statistics as mstat
+
+from idaes.core.util.misc import VarLikeExpression
+import idaes.logger as idaeslog
+
+
+
+_constR = 8.3145 * pyo.units.J / pyo.units.mol / pyo.units.K  # or Pa*m3/K/mol
+_constF = 96485 * pyo.units.coulomb / pyo.units.mol
+_scaling_factor_pressure = 1E-4
+_scaling_factor_temperature_thermo = 1E-2
+_scaling_factor_temperature_channel = 1E2
+_scaling_factor_temperature_conduction = 1E2
+_scaling_factor_concentration = 1
+_scaling_factor_Dconcentration = 10
+_scaling_factor_energy = 1e-3
+_safe_log_eps = 1e-9
+_safe_sqrt_eps = 1e-9
+_Tmax = None#900+273.15+300
+
+class CV_Bound(enum.Enum):
+    EXTRAPOLATE = 1
+    NODE_VALUE = 2
+
+
+class CV_Interpolation(enum.Enum):
+    UDS = 1  # Upwind difference scheme, exit face same as node center
+    CDS = 2  # Linear interpolation from upstream and downstream node centers
+    QUICK = 3  # Quadratic upwind interpolation, quadratic from two upwind
+    # centers and one downwind
+
+
+def _set_default_factor(c, s):
+    for i in c:
+        if iscale.get_scaling_factor(c[i]) is None:
+            iscale.set_scaling_factor(c[i], s)
+def _set_scaling_factor_if_none(c,s):
+    if iscale.get_scaling_factor(c) is None:
+        iscale.set_scaling_factor(c, s)
+
+def _set_and_get_scaling_factor(c,s):
+    _set_scaling_factor_if_none(c,s)
+    return iscale.get_scaling_factor(c)
+
+def _set_if_unfixed(v, val):
+    if not v.fixed:
+        v.set_value(pyo.value(val))
+        
+def _create_if_none(blk, var, idx_set, units):
+    attr = getattr(blk.config,var)
+    if attr is None:
+        if idx_set is None:
+            setattr(blk, var, 
+                    pyo.Var(
+                        initialize=0, 
+                        units=units,
+                        )
+            )
+        else:
+            setattr(blk, var, 
+                    pyo.Var(
+                        *idx_set,
+                        initialize=0, 
+                        units=units,
+                        )
+            )
+    else:
+        setattr(blk, var, pyo.Reference(attr))
+
+def _init_dof_error(blk):
+    if not mstat.degrees_of_freedom(blk) == 0:
+        raise InitializationError(
+            f"{blk.name} encountered a nonsquare optimization problem during "
```

Comment:
Maybe just "nonsquare problem."  I guess if you expect it to be square, it's not optimization.